### ISSUE TITLE
Proposal: Reset LiveSocket to disconnected if one of the hooks `mounted` function crashes

### ISSUE THREAD

**Issue by clemensm** [phoenixframework/phoenix_live_view#1388]
Hi everyone

I've just now had to fix a bug in one of my apps where LiveView would hang in an invalid/unusable state that was really hard to diagnose, and I think it would be helpfull if we could change the behaviour of the LiveView JS code slightly so that this is easier to catch. 

What happened is the following:

1. I've called the `connect()` function in my `app.js` code, connection is successfully established.
2. In the JS code the root Views are being joined
3. During the join the `mounted` callbacks are executed, sequentially, before the `phx-disconnected` flag is being cleared on the LiveView
4. I had a bug in one of the `mounted` callbacks that threw an error (that was sadly drowned out in the logs, my error obviously)
5. Thus the `phx-disconnected` flag never got cleared

Now when interacting with the app I had the following situation: I could still trigger changes (e.g. a `phx-click`) which would be send to the server. The server would reply as expected, sending a diff that should be merged. However after checking the JS code of the LiveView, I realized that the diff was never applied, because the function `isJoinPending()` was still returning `true`, and thus the diff was pushed onto the stack of pending diffs:

```
update(diff, events){
    if(this.isJoinPending() || this.liveSocket.hasPendingLink()){
      return this.pendingDiffs.push({diff, events})
    }
```

So long story short, would it be possible to catch errors in the `applyJoinPatch` function where the `mounted` callbacks are being executed and to reset the connection/report an error? So something like this:

```
applyJoinPatch(live_patch, html, events){
    try {
    DOM.all(this.el, `[${this.binding(PHX_HOOK)}], [data-phx-${PHX_HOOK}]`, hookEl => {
      let hook = this.addHook(hookEl)
      if(hook){ hook.__mounted() }
    })
    } catch (e) {
       console.error(e);
       console.error("Entering invalid state, disconnecting again because one of your hooks has thrown an execption");
       this.connection.disconnect();
    }
```

Thanks for your feedback and have a nice weekend 😄

**Comment by chrismccord** [comment]
Since live views are isolated, one crashing does not take out the concrete connection or other root live views, by design. I understand this was frustrating, but the solution would be to fix the bug on mount :) 
Thanks!
